### PR TITLE
Refactor apply tool to use HDF5Merger, read data only once

### LINKED ISSUE #2130: Allow dropping dl1 and telescope-wise information in `apply-models` or add tool for that.
**Please describe the use case that requires this feature.**
Currently, `apply-models` just copies the input file, with everything inside it.

We should add options – or a second tool – to only copy subsets, e.g. the metadata + instrument + event/subarray data

### LINKED ISSUE #2248: FeatureGenerator modifies and returns input table
Currently, the `FeatureGenerator` modifies and returns the input table.

This is a counter-intuitive API.
We should either:

* have it mutate the input and not return
* Create a new table and return that

Part of why I thought we need #2240 was that the reconstructors modify the input table, which should probably not be the case.

A couple of solutions I came up with:

* functional approach: return new table with added columns
* mutating approach: remove the return, allow the override

Both have significant drawbacks: I think it is never a good idea to mutate input in that way (reconstructors.predict_table should produce a prediction, not have side effects on the input table). But creating a full copy will also use much more memory.

I think a good solution would be to return a table only containing the new features and havng the `table_to_X` method of the sklearn reconstructures collect features from both the input table and the generated features table to keep both cleanly separated.

### LINKED ISSUE #2250: TableLoader breaks event order of telescope events when joining observation information
When joining the by-observation-block info, tables are sorted by obs_id and obs_id is converted to float

### PR DESCRIPTION
**Pull request by maxnoe**
Refactor ctapipe-apply models
    
* The tool now uses the HDF5Merger component for copying data from
the input file, which makes it possible to select what data should
be copied.

* Data is only read once in chunks from the input files and all
reconstructors are applied in order to each of the chunks.

For this, it was needed to not modify the input table in the feature generator, which I opted to do by the shallow copy approach as discussed here: #2248 

Fixes #2248 
Fixes #2130
Fixes #2250

### CODE REVIEW

File: ctapipe/tools/apply_models.py [hunk 1]
```diff
@@ -100,84 +138,72 @@ def setup(self):
...
 
-                    write_table(
-                        table[output_columns],
-                        self.output_path,
-                        f"/dl2/event/telescope/{prop}/{prefix}/tel_{tel_id:03d}",
-                        append=True,
-                    )
-                tel_tables.append(table)
+            if tel not in reconstructor._models:
+                self.log.warning(
```
Comment (LukasNickel) [inline ctapipe/tools/apply_models.py:172]: Should this be a warning or raise an error?
I am not entirely sure yet what I would expect to happen if my reconstructor has not been trained for the complete array.

Comment (maxnoe) [reply ctapipe/tools/apply_models.py:172]: It should probably be an error, yes. Will have to check how that works with our test files though.

**Review by kosack — APPROVED** [review]

**Comment by maxnoe** [comment]
When running this using the test_ctapipe_ml repo, I get warnings from table loader that for some columns the table order doesn't match, which I think means that we are writing out predictions in the wrong order. I have to investigate why and why the tests do not catch that.

**Comment by maxnoe** [comment]
The single telescope predictions are in the wrong order. and the obs-id for some reason is a float.

Will investigate and add tests.

**Comment by maxnoe** [comment]
Okay, issue in tableloader found and fixed

**Review by LukasNickel — APPROVED** [review]

File: ctapipe/tools/apply_models.py (hunk 1 above)
Comment (maxnoe) [reply ctapipe/tools/apply_models.py:174]: Since this is not a change to before, let's discuss in an issue

**Review by kosack — APPROVED** [review]

**Pull request merged**